src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -24,162 +25,252 @@ const (
+		_applyError(inTree, exitcode.InsufficientFunds_System, nil, nil, MinerPenaltySpec_ChargeUsedGas)
+		return
 	}
 
 	// make sure this is the right message order for fromActor
 	// (this is protection against replay attacks, and useful sequencing)
 	if message.CallSeqNum() != fromActor.CallSeqNum()+1 {
-		return inTree, _applyError(exitcode.InvalidCallSeqNum)
+		_applyError(inTree, exitcode.InvalidCallSeqNum, nil, nil, MinerPenaltySpec_ChargeUsedGas)
+		return
 	}
 
 	// WithActorForAddress may create new account actors
-	compTree, toActor = compTree.Impl().WithActorForAddress(message.To())
+	compTreePreSend := inTree
+	compTreePreSend, toActor := compTreePreSend.Impl().WithActorForAddress(message.To())
 	if toActor == nil {
-		return inTree, _applyError(exitcode.ActorNotFound)
+		// Execution error; receiver actor does not exist (and could not be implicitly created)
+		// at time of message execution.
+		_applyError(inTree, exitcode.ActorNotFound,
+			message.From().Ref(), minerAddr.Ref(), MinerPenaltySpec_None)
+		return
+	}
+
+	// Deduct gas limit funds from sender first.
+	// (This should always succeed, due to the sender balance check above.)
+	compTreePreSend = _withTransferFundsAssert(
+		compTreePreSend, message.From(), addr.BurntFundsActorAddr, gasLimitCost)
+
+	sendRet, compTreePostSend := _applyMessageInternal(compTreePreSend, message, vmiGasRemaining, minerAddr)
+
+	ok = _vmiUseGas(sendRet.GasUsed())
+	if !ok {
+		panic("Interpreter error: runtime execution used more gas than provided")
+	}
+
+	ok = _vmiUseGas(gascost.OnChainReturnValue(sendRet.ReturnValue()))
+	if !ok {
+		// In compTreePreSend payment for the gas limit has been transferred to burnt funds actor.
+		// When the return value serialization overflows the gas limit, pay the full balance to the miner,
+		// refunding none to the sender (even though the offending return-value won't go on-chain), and then fail.
+		_applyError(compTreePreSend, exitcode.OutOfGas,
+			addr.BurntFundsActorAddr.Ref(), minerAddr.Ref(), MinerPenaltySpec_None)
+		return
+	}
+
+	compTreeRet := compTreePreSend
+	if sendRet.ExitCode().AllowsStateUpdate() {
+		compTreeRet = compTreePostSend
 	}
 
-	// deduct maximum expenditure gas funds first
-	compTree = _withTransferFundsAssert(compTree, message.From(), addr.BurntFundsActorAddr, maxGasCost)
+	// Refund unused gas to sender.
+	refundGas := vmiGasRemaining
+	compTreeRet = _withTransferFundsAssert(
+		compTreeRet,
+		addr.BurntFundsActorAddr,
+		message.From(),
+		_gasToFIL(refundGas, message.GasPrice()),
+	)
+
+	_applyReturn(
+		compTreeRet, vmr.InvocOutput_Make(sendRet.ReturnValue()), sendRet.ExitCode(),
+		addr.BurntFundsActorAddr.Ref(), minerAddr.Ref(), MinerPenaltySpec_None)
+	return
+}
+
+func _applyMessageBuiltinAssert(tree st.StateTree, message msg.UnsignedMessage, topLevelBlockWinner addr.Address) st.StateTree {
+	gasRemainingInit := msg.GasAmount_SentinelUnlimited()
+	Assert(gasRemainingInit.Equals(message.GasLimit()))
```

Comment:
Just use message.GasLimit? I don't understand why a sentinel or "unlimited" is involved, this seems to be duplicating a concept that is already expressed at the builtin message construction.

Reply:
This is due to the fact that it's a system invocation and we should never allow it to run out of gas (moreover, it's not clear who we would charge the gas to, if it did).

Reply:
Hard disagree. We have to put an actual number into the real computer! There should be a real limit and our blockchain should halt if it's exceeded (it's gunna halt anyway), and we can fix it at the social consensus layer of humans.

Please pick a number after which "unlimited" runs out, preferably by restoring the constants I had at the top of this file. It can be large enough to be practically "never", but you can't IMPL_TODO explicitly dodging this.